### DISCUSSION TITLE
[control mode] how to get notified of window index change?

### DISCUSSION THREAD

**Discussion by laur89** [tmux/tmux#4230] [General]
I'm working on a script that keeps state of some very basic tmux data, such as attached clients, their sessions & windows. To update this state it's consuming various notifications via the control mode.

However when re-ordering windows within a given session, the only notification received is `%session-window-changed`, which contains no index data. Same notification is also fired whenever viewed/active windows are changed in the session.

Or am I approaching it in too high detail and should just request `list-windows` with a format containing all the needed data with each and every notification, effectively refreshing the whole state snapshot as opposed to trying to keep it up-to-date incrementally?

---

Unrelated to window index notification, but there are other notifications that feel like should be there. One that comes to mind is when a new client is started/attached, only notification fired is `%client-session-changed`

**Comment by nicm** [comment]
You will probably need to issue `list-windows` or `display-message` to get the window info. We can make changes to the arguments or add new notifications, just make sure it doesn't break iTerm2...

**Reply by laur89** [reply]
Thanks. So I understand new event definition for window index change wouldn't be out of the question?

As for `display-message` - reading the docs I don't quite understand how it would help me in this particular case. What am I missing here?

**Reply by nicm** [reply]
Depends what it is, if there is an existing notify then can add it to that but make sure it won't break iTerm2.

You know the window ID so you can use display-message to get whatever other info you need about it.

**Reply by laur89** [reply]
For future searchers, this is how to get e.g. window index via display-message:
```sh
$ tmux display-message -t @3 -p '#{window_index}'
```

**Comment by laur89** [comment]
Now that I think of it, modifying existing notification arguments makes no sense, as change of window index will very likely cause change in other windows' indices as well. Which means window move would trigger number of TBD notifications (e.g. `window-index-changed`).

It is however unfortunate that some window moves wouldn't trigger any notifications whatsoever, e.g. `swap-window  -t +1` without the `-d` option.

**Comment by laur89** [comment]
@nicm trying to add a new notification for window reorder/index change event; would you be so kind and point me to roughly right place where to look at?

Naively thought it's at [session_renumber_windows()](https://github.com/tmux/tmux/blob/563ed053539feeddd588d7502e3387a5fa254f0b/session.c#L703), but this part is not even called by say [cmd_swap_window_exec()](https://github.com/tmux/tmux/blob/563ed053539feeddd588d7502e3387a5fa254f0b/cmd-swap-window.c#L46). In fact I can't comprehend currently where the swap flow causes indices to be updated.

---

As for the new notification, I'm thinking there are 2 somewhat reasonable options:
1. a new `%window-order-changed  @win-id  index` message fired per window. This could probably use the existing `notify_winlink()`, eg `notify_winlink("window-order-changed", wl_new);`
2. a new `%session-windows-reordered  $sess-id  @win1-id  win1-idx ... @winN-id  winN-idx`

Am I on a reasonable path here?

**Reply by nicm** [reply]
I think you will need to trigger it anywhere the window indexes can be moved around, so `swap-window`, `move-window`, `renumber-windows`/`move-window -r`, `new-window -a`.

I think I would make it `window-moved` and `session-windows-moved` probably.
